### PR TITLE
Fix fragment handling in HttpClient

### PR DESCRIPTION
**Pull request by stephentoub**
SocketsHttpHandler isn't sending fragments, nor is it properly inheriting the fragment from the original request URI into the redirect location URI when the original URI had one and the redirect URI did not, even though RFC 7231 says it must.  This commit fixes that for SocketsHttpHandler.

WinHttpHandler also isn't handling this inheritance according to the RFC. It appears that the logic for WinHttpHandler would actually need to be changed in WINHTTP itself, or else WinHttpHandler would need to be changed to do the redirects itself.

Neither CurlHandler or NetFxHandler send fragments at all.

This commit also fixes the test to correctly compare the expected and actual Uris... apparently Uri equality doesn't factor in fragments, so they're first converted to strings.  It also updates the test to also validate that the server received the URI with the fragment included.

Closes https://github.com/dotnet/corefx/issues/27305
cc: @geoffkizer, @davidsh, @wfurt, @rmkerr

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 1]
```diff
@@ -912,40 +912,77 @@ public void GetAsync_ServerNeedsAuthAndNoCredential_StatusCodeUnauthorized()
...
-                    Task secondRequest = origServer.AcceptConnectionSendResponseAndCloseAsync();
+                    // Receive the second request.
+                    Task<List<string>> secondRequest = origServer.AcceptConnectionSendResponseAndCloseAsync();
                     await TestHelper.WhenAllCompletedOrAnyFailed(secondRequest, getResponse);
 
+                    // Make sure the server received the second request for the right Uri.
+                    Assert.NotEmpty(secondRequest.Result);
+                    string[] statusLineParts = secondRequest.Result[0].Split(' ');
+                    Assert.Equal(3, statusLineParts.Length);
+                    Assert.Equal(expectedUrl.GetComponents(UriComponents.PathAndQuery | UriComponents.Fragment, UriFormat.SafeUnescaped), statusLineParts[1]);
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Nit: shouldn't matter for this test, but shouldn't this technically use UriFormat.Escaped?

Comment (geoffkizer) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: LGTM

**Comment by geoffkizer** [comment]
LGTM, one minor nit above

**Comment by stephentoub** [comment]
@davidsh, why did you remove the System.Net.Http label?  This applies to all handlers.  There's a fix in SocketsHttpHandler, but the test is for all of them, and the others all suffer from the same or similar issues.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Windows x86 Release Build please (https://github.com/dotnet/corefx/issues/27363)

**Comment by davidsh** [comment]
>@davidsh, why did you remove the System.Net.Http label? This applies to all handlers. There's a fix in SocketsHttpHandler, but the test is for all of them, and the others all suffer from the same or similar issues.

I based the label on where the produce code changes were made primarily which was SocketsHttpHandler.  Doing this helps group all the PRs related to changes in SocketNetHttpHandler.

I don't have a strong opinion on this, so you want, the label can be changed to just the generic "Http" lable.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please please

**Pull request merged**